File: android-app/app/src/main/java/com/example/webrtcstreamer/webrtc/WebRTCClient.java
package com.example.webrtcstreamer.webrtc;

import android.app.Application;
import android.content.Context;
import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;
import org.webrtc.AudioSource;
import org.webrtc.AudioTrack;
import org.webrtc.Camera2Enumerator;
import org.webrtc.CameraVideoCapturer;
import org.webrtc.DataChannel;
import org.webrtc.DefaultVideoDecoderFactory;
import org.webrtc.DefaultVideoEncoderFactory;
import org.webrtc.EglBase;
import org.webrtc.IceCandidate;
import org.webrtc.MediaConstraints;
import org.webrtc.MediaStream;
import org.webrtc.PeerConnection;
import org.webrtc.PeerConnectionFactory;
import org.webrtc.RtpReceiver;
import org.webrtc.SdpObserver;
import org.webrtc.SessionDescription;
import org.webrtc.SurfaceTextureHelper;
import org.webrtc.SurfaceViewRenderer;
import org.webrtc.VideoSource;
import org.webrtc.VideoTrack;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class WebRTCClient {
    private static final String TAG = "WebRTCClient";

    private final Application application;
    private final PeerConnectionObserver observer;

    public final EglBase eglBase;
    private final PeerConnectionFactory peerConnectionFactory;
    private PeerConnection peerConnection;
    private VideoTrack localVideoTrack;
    private AudioTrack localAudioTrack;
    private CameraVideoCapturer videoCapturer;

    public interface OnIceCandidateListener {
        void onIceCandidate(String candidateJson);
    }

    public OnIceCandidateListener onIceCandidate;

    public interface Callback<T> {
        void onSuccess(T result);
        void onError(String error);
    }

    public static class PeerConnectionObserver {
        public void onStreamReady() {}
        public void onStreamError(String error) {}
    }

    public WebRTCClient(Application application, PeerConnectionObserver observer) {
        this.application = application;
        this.observer = observer;
        this.eglBase = EglBase.create();
        
        initPeerConnectionFactory(application);
        this.peerConnectionFactory = createPeerConnectionFactory();
    }

    private void initPeerConnectionFactory(Context context) {
        PeerConnectionFactory.InitializationOptions options = PeerConnectionFactory.InitializationOptions.builder(context)
                .setEnableInternalTracer(true)
                .createInitializationOptions();
        PeerConnectionFactory.initialize(options);
    }

    private PeerConnectionFactory createPeerConnectionFactory() {
        PeerConnectionFactory.Options options = new PeerConnectionFactory.Options();
        options.disableEncryption = false;
        options.disableNetworkMonitor = false;

        return PeerConnectionFactory.builder()
                .setVideoDecoderFactory(new DefaultVideoDecoderFactory(eglBase.getEglBaseContext()))
                .setVideoEncoderFactory(new DefaultVideoEncoderFactory(eglBase.getEglBaseContext(), true, true))
                .setOptions(options)
                .createPeerConnectionFactory();
    }

    public void startLocalVideoCapture(SurfaceViewRenderer localView) {
        try {
            Log.d(TAG, "Starting video capture...");
            SurfaceTextureHelper surfaceTextureHelper = SurfaceTextureHelper.create(
                    Thread.currentThread().getName(),
                    eglBase.getEglBaseContext()
            );

            Log.d(TAG, "Creating camera capturer...");
            videoCapturer = createCameraCapturer();
            Log.d(TAG, "Camera capturer created: " + videoCapturer);

            VideoSource videoSource = peerConnectionFactory.createVideoSource(videoCapturer.isScreencast());
            Log.d(TAG, "Video source created");

            videoCapturer.initialize(surfaceTextureHelper, application, videoSource.getCapturerObserver());
            Log.d(TAG, "Camera initialized");

            videoCapturer.startCapture(640, 480, 24);
            Log.d(TAG, "Camera capture started at 640x480@24fps");

            localVideoTrack = peerConnectionFactory.createVideoTrack("local_video_track", videoSource);
            localVideoTrack.addSink(localView);
            Log.d(TAG, "Video track created and sink added");

            // Create audio track
            AudioSource audioSource = peerConnectionFactory.createAudioSource(new MediaConstraints());
            localAudioTrack = peerConnectionFactory.createAudioTrack("local_audio_track", audioSource);
            Log.d(TAG, "Audio track created");

            observer.onStreamReady();
        } catch (Exception e) {
            Log.e(TAG, "Failed to start video capture", e);
            observer.onStreamError("Camera error: " + e.getMessage());
        }
    }

    private CameraVideoCapturer createCameraCapturer() {
        Camera2Enumerator enumerator = new Camera2Enumerator(application);
        String[] deviceNames = enumerator.getDeviceNames();
        
        Log.d(TAG, "Found " + deviceNames.length + " cameras");

        // Try BACK camera first
        for (String deviceName : deviceNames) {
            if (enumerator.isBackFacing(deviceName)) {
                CameraVideoCapturer capturer = enumerator.createCapturer(deviceName, null);
                if (capturer != null) {
                    Log.d(TAG, "Using back camera: " + deviceName);
                    return capturer;
                }
            }
        }

        // Try FRONT camera
        for (String deviceName : deviceNames) {
            if (enumerator.isFrontFacing(deviceName)) {
                CameraVideoCapturer capturer = enumerator.createCapturer(deviceName, null);
                if (capturer != null) {
                    Log.d(TAG, "Using front camera: " + deviceName);
                    return capturer;
                }
            }
        }

        throw new RuntimeException("No camera found on device");
    }

    public PeerConnection createPeerConnection() {
        List<PeerConnection.IceServer> iceServers = new ArrayList<>();
        iceServers.add(PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer());
        iceServers.add(PeerConnection.IceServer.builder("turn:13.234.129.181:3478")
                .setUsername("dome")
                .setPassword("domepass123")
                .createIceServer());

        PeerConnection.RTCConfiguration rtcConfig = new PeerConnection.RTCConfiguration(iceServers);
        rtcConfig.bundlePolicy = PeerConnection.BundlePolicy.MAXBUNDLE;
        rtcConfig.rtcpMuxPolicy = PeerConnection.RtcpMuxPolicy.REQUIRE;
        rtcConfig.tcpCandidatePolicy = PeerConnection.TcpCandidatePolicy.ENABLED;
        rtcConfig.continualGatheringPolicy = PeerConnection.ContinualGatheringPolicy.GATHER_CONTINUALLY;
        rtcConfig.keyType = PeerConnection.KeyType.ECDSA;
        rtcConfig.sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN;

        peerConnection = peerConnectionFactory.createPeerConnection(rtcConfig, new PeerConnection.Observer() {
            @Override
            public void onIceCandidate(IceCandidate candidate) {
                if (candidate != null && onIceCandidate != null) {
                    try {
                        JSONObject json = new JSONObject();
                        json.put("candidate", candidate.sdp);
                        json.put("sdpMid", candidate.sdpMid);
                        json.put("sdpMLineIndex", candidate.sdpMLineIndex);
                        onIceCandidate.onIceCandidate(json.toString());
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }
            }

            @Override
            public void onDataChannel(DataChannel dataChannel) {}

            @Override
            public void onIceConnectionReceivingChange(boolean b) {}

            @Override
            public void onIceConnectionChange(PeerConnection.IceConnectionState iceConnectionState) {
                Log.d(TAG, "ICE connection state changed to: " + iceConnectionState);
            }

            @Override
            public void onIceGatheringChange(PeerConnection.IceGatheringState iceGatheringState) {
                Log.d(TAG, "ICE gathering state changed to: " + iceGatheringState);
            }

            @Override
            public void onAddStream(MediaStream mediaStream) {}

            @Override
            public void onSignalingChange(PeerConnection.SignalingState signalingState) {}

            @Override
            public void onIceCandidatesRemoved(IceCandidate[] iceCandidates) {}

            @Override
            public void onRemoveStream(MediaStream mediaStream) {}

            @Override
            public void onRenegotiationNeeded() {}

            @Override
            public void onAddTrack(RtpReceiver rtpReceiver, MediaStream[] mediaStreams) {}
        });

        if (localVideoTrack != null) {
            peerConnection.addTrack(localVideoTrack, Collections.singletonList("local_stream"));
        }
        if (localAudioTrack != null) {
            peerConnection.addTrack(localAudioTrack, Collections.singletonList("local_stream"));
        }

        return peerConnection;
    }

    public void createOffer(Callback<String> callback) {
        if (peerConnection != null) {
            peerConnection.close();
        }
        peerConnection = createPeerConnection(); // Returns null? No, implementation above returns object.

        if (peerConnection == null) {
            callback.onError("Failed to create peer connection");
            return;
        }

        MediaConstraints constraints = new MediaConstraints();
        constraints.mandatory.add(new MediaConstraints.KeyValuePair("OfferToReceiveAudio", "false"));
        constraints.mandatory.add(new MediaConstraints.KeyValuePair("OfferToReceiveVideo", "false"));

        peerConnection.createOffer(new SdpObserver() {
            @Override
            public void onCreateSuccess(SessionDescription sdp) {
                Log.d(TAG, "Offer created successfully");
                peerConnection.setLocalDescription(new SdpObserver() {
                    @Override
                    public void onCreateSuccess(SessionDescription sessionDescription) {}

                    @Override
                    public void onSetSuccess() {
                        Log.d(TAG, "Local description set successfully");
                        try {
                            JSONObject json = new JSONObject();
                            json.put("type", sdp.type.canonicalForm());
                            json.put("sdp", sdp.description);
                            callback.onSuccess(json.toString());
                        } catch (JSONException e) {
                            callback.onError(e.getMessage());
                        }
                    }

                    @Override
                    public void onCreateFailure(String s) {}

                    @Override
                    public void onSetFailure(String s) {
                        Log.e(TAG, "Failed to set local description: " + s);
                        callback.onError(s);
                    }
                }, sdp);
            }

            @Override
            public void onSetSuccess() {}

            @Override
            public void onCreateFailure(String s) {
                Log.e(TAG, "Failed to create offer: " + s);
                callback.onError(s);
            }

            @Override
            public void onSetFailure(String s) {}
        }, constraints);
    }

    public void setRemoteAnswer(String answerJson, Callback<Void> callback) {
        String answerSdp = answerJson;
        try {
            JSONObject json = new JSONObject(answerJson);
            answerSdp = json.getString("sdp");
        } catch (JSONException e) {
            // fallback to raw string
        }

        SessionDescription sdp = new SessionDescription(SessionDescription.Type.ANSWER, answerSdp);
        peerConnection.setRemoteDescription(new SdpObserver() {
            @Override
            public void onCreateSuccess(SessionDescription sessionDescription) {}

            @Override
            public void onSetSuccess() {
                Log.d(TAG, "setRemoteAnswer SUCCESS");
                callback.onSuccess(null);
            }

            @Override
            public void onCreateFailure(String s) {}

            @Override
            public void onSetFailure(String s) {
                Log.e(TAG, "setRemoteAnswer FAILED: " + s);
                callback.onError(s);
            }
        }, sdp);
    }

    public void addRemoteIceCandidate(String candidateJson) {
        try {
            JSONObject json = new JSONObject(candidateJson);
            String sdp = json.getString("candidate");
            String sdpMid = json.getString("sdpMid");
            int sdpMLineIndex = json.getInt("sdpMLineIndex");

            IceCandidate iceCandidate = new IceCandidate(sdpMid, sdpMLineIndex, sdp);
            if (peerConnection != null) {
                peerConnection.addIceCandidate(iceCandidate);
            }
        } catch (JSONException e) {
            observer.onStreamError("Failed to add ICE candidate: " + e.getMessage());
        }
    }

    public void close() {
        if (videoCapturer != null) {
            try {
                videoCapturer.stopCapture();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            videoCapturer.dispose();
            videoCapturer = null;
        }
        if (localVideoTrack != null) {
            localVideoTrack.dispose();
            localVideoTrack = null;
        }
        if (localAudioTrack != null) {
            localAudioTrack.dispose();
            localAudioTrack = null;
        }
        if (peerConnection != null) {
            peerConnection.close();
            peerConnection = null;
        }
        if (peerConnectionFactory != null) {
            peerConnectionFactory.dispose();
            // peerConnectionFactory = null; // final field
        }
        if (eglBase != null) {
            eglBase.release();
        }
    }
}

File: dome_cloud_deploy/android-app/app/src/main/java/com/example/webrtcstreamer/webrtc/WebRTCClient.kt
package com.example.webrtcstreamer.webrtc

import android.app.Application
import android.content.Context
import kotlinx.coroutines.suspendCancellableCoroutine
import org.webrtc.*
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class WebRTCClient(
    private val application: Application,
    private val observer: PeerConnectionObserver
) {
    
    val eglBase: EglBase = EglBase.create()
    private val peerConnectionFactory: PeerConnectionFactory
    private var peerConnection: PeerConnection? = null
    private var localVideoTrack: VideoTrack? = null
    private var localAudioTrack: AudioTrack? = null
    private var videoCapturer: CameraVideoCapturer? = null
    
    var onIceCandidate: ((String) -> Unit)? = null
    
    init {
        initPeerConnectionFactory(application)
        peerConnectionFactory = createPeerConnectionFactory()
    }
    
    private fun initPeerConnectionFactory(context: Context) {
        val options = PeerConnectionFactory.InitializationOptions.builder(context)
            .setEnableInternalTracer(true)
            .createInitializationOptions()
        PeerConnectionFactory.initialize(options)
    }
    
    private fun createPeerConnectionFactory(): PeerConnectionFactory {
        return PeerConnectionFactory.builder()
            .setVideoDecoderFactory(DefaultVideoDecoderFactory(eglBase.eglBaseContext))
            .setVideoEncoderFactory(
                DefaultVideoEncoderFactory(
                    eglBase.eglBaseContext,
                    true,
                    true
                )
            )
            .setOptions(PeerConnectionFactory.Options().apply {
                disableEncryption = false
                disableNetworkMonitor = false
            })
            .createPeerConnectionFactory()
    }
    
    fun startLocalVideoCapture(localView: SurfaceViewRenderer) {
        try {
            println("DEBUG: Starting video capture...")
            val surfaceTextureHelper = SurfaceTextureHelper.create(
                Thread.currentThread().name,
                eglBase.eglBaseContext
            )
            
            println("DEBUG: Creating camera capturer...")
            videoCapturer = createCameraCapturer()
            println("DEBUG: Camera capturer created: $videoCapturer")
            
            val videoSource = peerConnectionFactory.createVideoSource(videoCapturer!!.isScreencast)
            println("DEBUG: Video source created")
            
            videoCapturer!!.initialize(
                surfaceTextureHelper,
                application,
                videoSource.capturerObserver
            )
            println("DEBUG: Camera initialized")
            
            videoCapturer!!.startCapture(640, 480, 24)
            println("DEBUG: Camera capture started at 640x480@24fps")
            
            localVideoTrack = peerConnectionFactory.createVideoTrack("local_video_track", videoSource)
            localVideoTrack?.addSink(localView)
            println("DEBUG: Video track created and sink added")
            
            // Create audio track
            val audioSource = peerConnectionFactory.createAudioSource(MediaConstraints())
            localAudioTrack = peerConnectionFactory.createAudioTrack("local_audio_track", audioSource)
            println("DEBUG: Audio track created")
            
            observer.onStreamReady()
        } catch (e: Exception) {
            println("ERROR: Failed to start video capture: ${e.message}")
            e.printStackTrace()
            observer.onStreamError("Camera error: ${e.message}")
        }
    }
    
    private fun createCameraCapturer(): CameraVideoCapturer {
        val enumerator = Camera2Enumerator(application)
        
        val deviceNames = enumerator.deviceNames
        println("DEBUG: Found ${deviceNames.size} cameras: ${deviceNames.joinToString()}")
        
        // Try BACK camera first (usually more reliable)
        for (deviceName in deviceNames) {
            println("DEBUG: Checking camera: $deviceName, isBack=${enumerator.isBackFacing(deviceName)}")
            if (enumerator.isBackFacing(deviceName)) {
                val capturer = enumerator.createCapturer(deviceName, null)
                if (capturer != null) {
                    println("DEBUG: ✅ Using back camera: $deviceName")
                    return capturer
                }
            }
        }
        
        // If no back camera, try front camera
        for (deviceName in deviceNames) {
            println("DEBUG: Checking camera: $deviceName, isFront=${enumerator.isFrontFacing(deviceName)}")
            if (enumerator.isFrontFacing(deviceName)) {
                val capturer = enumerator.createCapturer(deviceName, null)
                if (capturer != null) {
                    println("DEBUG: ✅ Using front camera: $deviceName")
                    return capturer
                }
            }
        }
        
        throw RuntimeException("No camera found on device")
    }
    
    fun createPeerConnection(): PeerConnection? {
        val rtcConfig = PeerConnection.RTCConfiguration(
            listOf(
                PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer(),
                PeerConnection.IceServer.builder("stun:stun1.l.google.com:19302").createIceServer(),
                PeerConnection.IceServer.builder("stun:stun2.l.google.com:19302").createIceServer(),
                PeerConnection.IceServer.builder("turn:openrelay.metered.ca:80")
                    .setUsername("openrelayproject")
                    .setPassword("openrelayproject")
                    .createIceServer(),
                PeerConnection.IceServer.builder("turn:openrelay.metered.ca:443")
                    .setUsername("openrelayproject")
                    .setPassword("openrelayproject")
                    .createIceServer(),
                PeerConnection.IceServer.builder("turn:openrelay.metered.ca:443?transport=tcp")
                    .setUsername("openrelayproject")
                    .setPassword("openrelayproject")
                    .createIceServer()
            )
        ).apply {
            bundlePolicy = PeerConnection.BundlePolicy.MAXBUNDLE
            rtcpMuxPolicy = PeerConnection.RtcpMuxPolicy.REQUIRE
            tcpCandidatePolicy = PeerConnection.TcpCandidatePolicy.ENABLED
            iceTransportsType = PeerConnection.IceTransportsType.ALL
            continualGatheringPolicy = PeerConnection.ContinualGatheringPolicy.GATHER_CONTINUALLY
            keyType = PeerConnection.KeyType.ECDSA
        }
        
        peerConnection = peerConnectionFactory.createPeerConnection(
            rtcConfig,
            object : PeerConnection.Observer {
                override fun onIceCandidate(candidate: IceCandidate?) {
                    candidate?.let {
                        val candidateJson = """
                            {
                                "candidate": "${it.sdp}",
                                "sdpMid": "${it.sdpMid}",
                                "sdpMLineIndex": ${it.sdpMLineIndex}
                            }
                        """.trimIndent()
                        onIceCandidate?.invoke(candidateJson)
                    }
                }
                
                override fun onDataChannel(channel: DataChannel?) {}
                override fun onIceConnectionReceivingChange(receiving: Boolean) {
                    println("DEBUG: ICE connection receiving change: $receiving")
                }
                override fun onIceConnectionChange(newState: PeerConnection.IceConnectionState?) {
                    println("DEBUG: ICE connection state changed to: $newState")
                }
                override fun onIceGatheringChange(newState: PeerConnection.IceGatheringState?) {
                    println("DEBUG: ICE gathering state changed to: $newState")
                }
                override fun onIceCandidatesRemoved(candidates: Array<out IceCandidate>?) {}
                override fun onAddStream(stream: MediaStream?) {}
                override fun onSignalingChange(newState: PeerConnection.SignalingState?) {
                    println("DEBUG: Signaling state changed to: $newState")
                }
                override fun onRemoveStream(stream: MediaStream?) {}
                override fun onRenegotiationNeeded() {
                    println("DEBUG: Renegotiation needed")
                }
                override fun onAddTrack(receiver: RtpReceiver?, streams: Array<out MediaStream>?) {}
            }
        )
        
        // Add local tracks to peer connection
        localVideoTrack?.let { peerConnection?.addTrack(it, listOf("local_stream")) }
        localAudioTrack?.let { peerConnection?.addTrack(it, listOf("local_stream")) }
        
        return peerConnection
    }
    
    suspend fun createOffer(): String {
        // Close existing peer connection if any
        peerConnection?.close()
        
        // Create new peer connection
        val pc = createPeerConnection() ?: throw RuntimeException("Failed to create peer connection")
        
        println("DEBUG: Created peer connection, tracks added: video=${localVideoTrack != null}, audio=${localAudioTrack != null}")
        
        val constraints = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "false"))
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "false"))
        }
        
        return suspendCancellableCoroutine { continuation ->
            pc.createOffer(object : SdpObserver {
                override fun onCreateSuccess(sdp: SessionDescription?) {
                    println("DEBUG: Offer created successfully")
                    pc.setLocalDescription(object : SdpObserver {
                        override fun onSetSuccess() {
                            println("DEBUG: Local description set successfully")
                            sdp?.let {
                                // Return JSON with type and sdp
                                val jsonOffer = """{"type":"${it.type.canonicalForm()}","sdp":"${it.description}"}"""
                                println("DEBUG: Sending offer (length: ${jsonOffer.length})")
                                continuation.resume(jsonOffer)
                            }
                        }
                        override fun onSetFailure(error: String?) {
                            println("ERROR: Failed to set local description: $error")
                            continuation.resumeWithException(RuntimeException(error))
                        }
                        override fun onCreateSuccess(p0: SessionDescription?) {}
                        override fun onCreateFailure(p0: String?) {}
                    }, sdp)
                }
                
                override fun onCreateFailure(error: String?) {
                    println("ERROR: Failed to create offer: $error")
                    continuation.resumeWithException(RuntimeException(error))
                }
                
                override fun onSetSuccess() {}
                override fun onSetFailure(error: String?) {}
            }, constraints)
        }
    }
    
    suspend fun setRemoteAnswer(answerJson: String) {
        println("DEBUG: setRemoteAnswer called with: ${answerJson.take(100)}")
        // Parse answer JSON to extract SDP
        val answerSdp = try {
            val json = org.json.JSONObject(answerJson)
            json.getString("sdp")
        } catch (e: Exception) {
            println("DEBUG: Failed to parse as JSON, using raw string")
            // Fallback: if it's already just the SDP string
            answerJson
        }
        
        println("DEBUG: Setting remote answer SDP (length: ${answerSdp.length})")
        val sdp = SessionDescription(SessionDescription.Type.ANSWER, answerSdp)
        
        return suspendCancellableCoroutine { continuation ->
            peerConnection?.setRemoteDescription(object : SdpObserver {
                override fun onSetSuccess() {
                    println("DEBUG: setRemoteAnswer SUCCESS!")
                    continuation.resume(Unit)
                }
                
                override fun onSetFailure(error: String?) {
                    println("ERROR: setRemoteAnswer FAILED: $error")
                    continuation.resumeWithException(RuntimeException(error))
                }
                
                override fun onCreateSuccess(p0: SessionDescription?) {}
                override fun onCreateFailure(p0: String?) {}
            }, sdp)
        }
    }
    
    suspend fun addRemoteIceCandidate(candidateJson: String) {
        try {
            // Parse the ICE candidate JSON
            val json = org.json.JSONObject(candidateJson)
            val sdp = json.getString("candidate")
            val sdpMid = json.getString("sdpMid")
            val sdpMLineIndex = json.getInt("sdpMLineIndex")
            
            val iceCandidate = IceCandidate(sdpMid, sdpMLineIndex, sdp)
            peerConnection?.addIceCandidate(iceCandidate)
        } catch (e: Exception) {
            observer.onStreamError("Failed to add ICE candidate: ${e.message}")
        }
    }
    
    fun close() {
        videoCapturer?.stopCapture()
        videoCapturer?.dispose()
        localVideoTrack?.dispose()
        localAudioTrack?.dispose()
        peerConnection?.close()
        peerConnectionFactory.dispose()
        eglBase.release()
    }
    
    open class PeerConnectionObserver {
        open fun onStreamReady() {}
        open fun onStreamError(error: String) {}
    }
}
